feat: import files to MFS

colinfruit opened the pull request:
This PR changes the upload target from `ipfs add` to MFS. Currently, it uses the default directory `ipfs-companion-uploads` with the user option to change this on the upload page. On successful upload, it will open the Web UI, at the directory used for the upload. 

closes #415 and closes #635

lidel commented on add-on/src/popup/quick-upload.js:
```diff
@@ -61,34 +59,28 @@ async function processFiles (state, emitter, files) {
+      parents: true
     }
-    let result
+    state.progress = `Uploading ${streams.length} files...`
```
Note to self: it looks like we are losing import progress report, but it never worked correctly over HTTP API (https://github.com/ipfs/js-ipfs-http-client/issues/842), so its ok to drop it for now.

lidel commented on add-on/src/lib/ipfs-companion.js:
```diff
@@ -787,6 +793,10 @@ module.exports = async function init () {
 
+    get openWebUiAtDirectory () {
+      return openWebUiAtDirectory
+    },
```
Ugh, another realization: we can't open webui if embedded js-ipfs is used in non-Brave browser.
We need to check if `state.nodeType === 'embedded'` and open resource at public gateway instead.

lidel replied:
Perhaps we could make this even more useful:
What if we add toggle "Open imported files in WebUI"?
Enabled by default, but when disabled it would open at Gateway (restoring old behavior).

colinfruit replied:
Sounds good. It would be great in subsequent issue to find a way of handling opening the WebUI when using embedded nodes. Has this been looked at before?

colinfruit commented on add-on/src/popup/quick-upload.js:
```diff
@@ -61,34 +59,28 @@ async function processFiles (state, emitter, files) {
-        throw new Error('Result of ipfs.add call is missing entries. This may be due to a bug in HTTP API similar to https://github.com/ipfs/go-ipfs/issues/5168')
-      }
-      await ipfsCompanion.uploadResultHandler({ result, openRootInNewTab: true })
+      const files = streams.map(stream => (ipfsCompanion.ipfs.files.write(`${uploadDir}${stream.path}`, stream.content, options)))
```
Thanks!

colinfruit commented:
Thanks, @lidel!

> Please change labels and variable names from "upload" to "import"

Makes sense. I changed all the user-facing labels this PR introduces from "upload" to "import". I did not change the variable names, because it feels inconsistent with the existing code. References to `upload` are everywhere, even the file name is `quick-upload.js`. If you want to make an issue, I'll be glad to remove `upload` references throughout the code base.

> Would it be possible to expose the name of default directory for uploads as an editable setting on Preferences screen?

Sure. Also, I removed it from the upload screen, because it seemed redundant. On the other hand, it provides useful functionality there as well. What do you think?

> I am bit worried what happens to /ipfs-companion-imports/ when user uploads 10th, 100th file. 

I'm looking into the best way to handle user-generated date symbols (`%Y-%m-%d`), and the best way to demonstrate this functionality to the user. What if they removed them, and then forgot what the symbols were, or weren't sure what they meant to begin with? Maybe a result field on the Preferences screen (and upload screen if we return the functionality) that show what directory will be used? Or an instruction in the description?

lidel commented:
Thank you @colinfruit! I am travelling this week, but will review this as soon I have some uninterrupted time.

lidel commented on add-on/_locales/en/messages.json:
```diff
@@ -427,6 +427,14 @@
     "description": "An option description for the log level (option_logNamespaces_description)"
   },
+  "option_uploadDir_title": {
+    "message": "MFS Import Directory",
```
We try to avoid "MFS" in user interfaces as it is a technical term and regular users don't know what it means.
```suggestion
    "message": "File Import Directory",
```

lidel commented on add-on/_locales/en/messages.json:
```diff
@@ -472,16 +480,20 @@
     "description": "Button on the share files page (quickUpload_options_show)"
   },
+  "quickUpload_options_uploadDir": {
+    "message": "MFS directory to store imported files",
```
```suggestion
    "message": "Path to store imported files",
```

lidel commented on add-on/_locales/en/messages.json:
```diff
@@ -472,16 +480,20 @@
-    "message": "Pin files so they are retained when performing garbage collection on your IPFS repo.",
-    "description": "Checkbox label on the share files page (quickUpload_options_pinUpload)"
+  "quickUpload_options_openViaWebUI": {
+    "message": "Open imported files in Web UI",
```
"imported files" is already in `quickUpload_options_uploadDir`, so we can make this one shorter or need to rephrase it
```suggestion
    "message": "Open in Web UI",
```

colinfruit commented on add-on/src/options/forms/experiments-form.js:
```diff
@@ -153,6 +154,26 @@ function experimentsForm ({
+          <input
+            id="uploadDir"
+            type="text"
+            pattern="^\/(.*)"
```
I added this before import happens in `quick-upload.js`, to account for the potential case of someone forgetting the trailing slash on the import page. It handles both cases, and prevents the user from having to guess that a trailing slash is necessary. https://github.com/ipfs-shipyard/ipfs-companion/pull/810/files#diff-9984375a8301f7832fa0da0a5aafd803R75

colinfruit commented:
> Let's add "File Import" section to Options page and move related settings there

colinfruit commented on add-on/src/popup/quick-upload.js:
```diff
@@ -81,20 +87,36 @@ async function processFiles (state, emitter, files) {
+    // unless and embedded node is in use (no access to web UI)
+    // in which case, open resource.
+    if (state.ipfsNodeType === 'embedded' || !state.openViaWebUI) {
+      await ipfsCompanion.uploadResultHandler({ result, openRootInNewTab: true })
```
Noticed this is not working as expected for me tonight. Maybe it's some weirdness with my connection; I'll take a look at this tomorrow.

colinfruit replied:
Same issue this morning, importing multiple files and opening them in a context menu hangs. However, it is reproducible on current master, so I won't deal with it in this PR.

colinfruit commented on add-on/src/popup/quick-upload.js:
```diff
@@ -134,6 +134,8 @@ function file2buffer (file) {
 
 function formatImportDirectory (path) {
   path = path.replace(/\/$|$/, '/')
+  path = path.replace(/(\/)\/+/g, '$1')
```
I added this just because it makes a nicer user experience. Files will appear in the Web UI with multiple slashes, despite being stored without them if multiple slashes are not removed on format.

lidel replied:
Thanks! Do you mind adding unit tests for `formatImportDirectory`, so we don't break this during future refactors?

For now it can be `test/functional/lib/quick-upload.test.js` (similar to existing `test/functional/lib/ipfs-path.test.js`)

lidel commented on add-on/src/popup/quick-upload.js:
```diff
@@ -61,17 +71,20 @@ async function processFiles (state, emitter, files) {
-      progress: (len) => progressHandler(len, totalSize, state, emitter),
-      wrapWithDirectory: wrapFlag,
-      pin: state.pinUpload
+      wrapWithDirectory: wrapFlag
```
The `ipfs.add` command  adds a low level pin by default.
We use MFS instead of low level pins, so we need to disable  it here:


```suggestion
      wrapWithDirectory: wrapFlag,
      pin: false // we use MFS for implicit pinning instead
```

colinfruit commented:
@lidel, fine to merge this. Thanks!

lidel commented:
@colinfruit any way to contact you privately? mind emailing me at `lidel<at>lidel.org`?
